lib/charms/opensearch/v0/opensearch_performance_profile.py

```diff
@@ -0,0 +1,214 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Represents the performance profile of the OpenSearch cluster.
+
+The main goals of this library is to provide a way to manage the performance
+profile of the OpenSearch cluster.
+
+There are two ways the charm can learn about its profile and when it changes:
+1) If this is the MAIN_ORCHESTRATOR: config-changed -> the user has switched the profile directly
+2) If not the MAIN_ORCHESTRATOR: peer-cluster-relation-changed -> the main orchestrator has
+                                 switched the profile
+
+The charm will then apply the profile and restart the OpenSearch service if needed.
+"""
+import logging
+import os
+
+import ops
+from charms.opensearch.v0.constants_charm import (
+    PERFORMANCE_PROFILE,
+    PeerClusterRelationName,
+)
+from charms.opensearch.v0.models import (
+    DeploymentType,
+    OpenSearchPerfProfile,
+    PerformanceType,
+)
+from charms.opensearch.v0.opensearch_exceptions import OpenSearchHttpError
+from charms.opensearch.v0.opensearch_internal_data import Scope
+from ops.framework import EventBase, EventSource
+
+# The unique Charmhub library identifier, never change it
+LIBID = "8b7aa39016e748ea908787df1d7fb089"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+
+logger = logging.getLogger(__name__)
+
+
+class _ApplyProfileTemplatesOpenSearch(EventBase):
+    """Attempt to acquire lock & restart OpenSearch.
+
+    The main reason to have a separate event, is to be able to wait for the cluster.
+    It defers otherwise and only defers the execution of this particular task.
+    """
+
+
+class OpenSearchPerformance(ops.Object):
+    """Base class for OpenSearch charms."""
+
+    _apply_profile_templates_event = EventSource(_ApplyProfileTemplatesOpenSearch)
+
+    def __init__(self, charm: ops.charm.CharmBase = None):
+        super().__init__(charm, None)
+        self.charm = charm
+        self.peers_data = self.charm.peers_data
+
+        self.framework.observe(
+            charm.on.install,
+            self._on_install,
+        )
+        self.framework.observe(
+            charm.on[PeerClusterRelationName].relation_changed,
+            self._on_peer_cluster_relation_changed,
+        )
+        self.framework.observe(
+            self._apply_profile_templates_event, self._on_apply_profile_templates
+        )
+        self._apply_profile_templates_has_been_called = False
+
+    def _on_install(self, event: EventBase):
```

Comment:
1. I need to make sure we have that set before the very first config-changed is called, from Juju docs: [install -> config-changed -> start](https://juju.is/docs/sdk/config-changed-event)
2. So, I put it on install because: (1) it goes back to encapsulating the internals of this component to itself; and (2) I don't want to have yet another option if you ask `opensearch_performance_profile.current`: I'd rather have it 
setting as PerformanceType only, not PerformanceType|None.

Now, that said, we can find another place given the constraints above!

Reply:
I understand the lifecycle, but we are mixing concerns here - install hook is about installing. 
Given you're now relying on the DeploymentDesc, there is no difference between this flow and the flow of setting roles at deploy time -- they're both taken into account in `new_cluster_setup`.

Reply:
Ack, it is also causing problems with upgrades. So, I am assigning it explicitly at the beginning of config-changed.